### PR TITLE
Add command infrastructure

### PR DESCRIPTION
**Pull request by neithernut**
This PR introduces an initial version of the command infrasructure.
Targets #188.

Please let me know if and how the documentation is not sufficient.

### CODE REVIEW

File: src/command/command.h [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+/**
+ * Command type
+ *
+ * This datatype represents a command, which may be invoked in a transaction.
+ */
+struct ws_command {
+    char const* name; //!< @public name of the command
+
+    enum {
+        regular, //!< it's a regulat command
```
Comment (TheNeikos) [inline src/command/command.h]: `regular`

File: src/command/command.h [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+ * There are two types of arguments: the ones passed directly and the ones which
+ * are already on the stack.
+ */
+struct argument {
+    enum {
+        indirect, //!< argument is given by position
+        direct //!< argument is given directly
+    } type; //!< @public type of argument (direct or indirect)
+    union {
+        ssize_t pos; //!< a position
```
Comment (matthiasbeyer) [inline src/command/command.h]: Why `ssize_t` - documentation missing here.

File: src/command/command.h [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+ */
+struct ws_command_args {
+    size_t num; //!< @public number of arguments
+    struct argument* vals; //!< @public arguments
+};
+
+/**
+ * Function type for command implementation
+ *
+ * The function
```
Comment (matthiasbeyer) [inline src/command/command.h]: Here's documentation missing.

File: src/command/command.h [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+typedef int (*ws_special_command_func)(struct ws_command_processor*,
+                                       struct ws_command_args*);
+
+/**
+ * Command type
+ *
+ * This datatype represents a command, which may be invoked in a transaction.
+ */
+struct ws_command {
+    char const* name; //!< @public name of the command
```
Comment (matthiasbeyer) [inline src/command/command.h]: Could this possibly be a `char const* const name`?

File: src/command/command.h [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+    enum {
+        regular, //!< it's a regular command
+        special //!< it's a special command (like jump)
+    } command_type; //!< @public type of the command
+
+    union {
+        ws_regular_command_func regular; //!< @public regular commands' callback
+        ws_special_command_func special; //!< @public special commands' callback
+    } func; //!< @public function callback
```
Comment (matthiasbeyer) [inline src/command/command.h:99]: Suggestion: The command type and the callback are directly related. Why not putting it into an own struct like so:

``` C
struct {
    enum { /* */ } type;
    union { /* */ } func;
} callback;
```

Comment (neithernut) [inline src/command/command.h:99]: If you look at the function definition, you'll notice that there is a lot of documentation, which I didn't want in the struct.

Comment (matthiasbeyer) [inline src/command/command.h:99]: That's not what I wanted to explain:

You actually can put the union and the enum above into an own scope, as they belong to eachother very closely.

Comment (neithernut) [inline src/command/command.h:99]: Suboptimal alignment.

Comment (matthiasbeyer) [inline src/command/command.h:99]: Are you sure about this? The additional struct doesn't add any data to the type, just a scope.

Comment (neithernut) [inline src/command/command.h:99]: The struct may align to `2*sizeof(void*)`.

File: src/command/command.h (hunk 2 above)
Comment (neithernut) [inline src/command/command.h]: I would do this in a more central place. Preferably in the module description.

Comment (matthiasbeyer) [inline src/command/command.h]: I'd prefer a note here. Documentation belongs to the code it documents, not somewhere else!

Comment (neithernut) [inline src/command/command.h]: The question is whether you get it, if I document it here.

Comment (matthiasbeyer) [inline src/command/command.h]: That's true. Then, a `@see` would be good.

**Comment by neithernut** [comment]
Any more anotations on this one?

File: src/command/command.h [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+
+#include <malloc.h>
+
+// forward declarations
+struct ws_command_processor;
+struct ws_value;
+
+/**
+ * A command agument
+ *
```
Comment (xVin) [inline src/command/command.h:39]: typo: argument

File: src/command/command.h [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+struct ws_command_processor;
+struct ws_value;
+
+/**
+ * A command argument
+ *
+ * There are two types of arguments: the ones passed directly and the ones which
+ * are already on the stack.
+ */
+struct argument {
```
Comment (matthiasbeyer) [inline src/command/command.h]: This does not fit our naming scheme here.

Comment (neithernut) [inline src/command/command.h]: Uh, forgot to prefix it when I pulled it out of another struct.

File: src/command/command.c [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+
+/*
+ *
+ * Interface implementation
+ *
+ */
+
+struct ws_command*
+ws_command_get(
```
Comment (matthiasbeyer) [inline src/command/command.c:38]: Documentation missing. I don't know what this function is meant to be, but I guess that "get" is a bit inappropriate here, but I don't know what this function is for...

Comment (neithernut) [inline src/command/command.c:38]: The function is documented in the header.

File: src/command/command.h [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+
+/**
+ * Command arguments
+ * 
+ * This struct represents command arguments applied to a command as part of a
+ * statement.
+ */
+struct ws_command_args {
+    size_t num; //!< @public number of arguments
+    struct argument* vals; //!< @public arguments
```
Comment (matthiasbeyer) [inline src/command/command.h]: Mustn't this be renamed to `ws_argument`?

Comment (neithernut) [inline src/command/command.h]: Er... yep.

File: src/command/command.c (hunk 8 above)
Comment (matthiasbeyer) [inline src/command/command.c:38]: Whoops.

**Comment by neithernut** [comment]
Any more annotations? I'm currently waiting for this to be merged.

**Comment by matthiasbeyer** [comment]
Nope, no annotations from me. ACK.

**Pull request merged**